pkg/util/grpc/encoding/snappy/snappy.go

```diff
@@ -0,0 +1,83 @@
+package snappy
+
+import (
+	"io"
+	"sync"
+
+	"github.com/golang/snappy"
+	"google.golang.org/grpc/encoding"
+)
+
+// Name is the name registered for the snappy compressor.
+const Name = "snappy"
+
+func init() {
+	encoding.RegisterCompressor(newCompressor())
+}
+
+type compressor struct {
+	writersPool sync.Pool
+	readersPool sync.Pool
+}
+
+func newCompressor() *compressor {
+	c := &compressor{}
+	c.readersPool = sync.Pool{
+		New: func() interface{} {
+			return snappy.NewReader(nil)
+		},
+	}
+	c.writersPool = sync.Pool{
+		New: func() interface{} {
+			return snappy.NewBufferedWriter(nil)
+		},
+	}
+	return c
+}
+
+func (c *compressor) Name() string {
+	return Name
+}
+
+func (c *compressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	wr := c.writersPool.Get().(*snappy.Writer)
+	wr.Reset(w)
+	return writeCloser{wr, &c.writersPool}, nil
+}
+
+func (c *compressor) Decompress(r io.Reader) (io.Reader, error) {
+	dr := c.readersPool.Get().(*snappy.Reader)
+	dr.Reset(r)
+	return reader{dr, &c.readersPool}, nil
+}
+
+type writeCloser struct {
+	*snappy.Writer
```

Comment:
Should this be a named `writer *snappy.Writer` too, according to the previous feedback?

Reply:
If we use `writer *snappy.Writer`, we need to add `func (* writeCloser)Write(p []byte) (n int, err error)`.
In additional, this type is private, no one can access it outside.

Reply:
Type is private, but methods it implements are not. Right now it exposes 4 methods (`Reset`, `Write`, `Flush`, `Close`), which is not clear from the code, and it's clear whether it was intended or not.

Reply:
(`Close` is reimplemented, so that one is clear)